portal/flatpak-portal.c

```diff
@@ -247,10 +251,188 @@ typedef struct
+
+      if (g_str_has_prefix (line, "NSpid:"))
+        {
+          g_auto(GStrv) fields = NULL;
+          guint nfields = 0;
+          char *endptr = NULL;
+
+          fields = g_strsplit (line, "\t", -1);
+          nfields = g_strv_length (fields);
+          if (nfields < 3)
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                           "NSpid line has too few fields: %s", line);
+              return 0;
+            }
+
+          /* The second to last PID namespace is the one that spawned this process */
+          relative_pid = strtol (fields[nfields - 2], &endptr, 10);
+          if (*endptr)
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                           "Invalid parent-relative PID in NSpid line: %s", line);
+              return 0;
+            }
+
+          return relative_pid;
+        }
+    }
+
+  if (*error == NULL)
+    /* EOF was reached while reading the file */
+    g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND, "NSpid not found");
+
+  return 0;
+}
+
+static int
+check_child_pid_status (void *user_data)
+{
+  /* Stores a sequence of the time interval to use until the child PID is checked again.
+     In general from testing, bwrapinfo is never ready before 25ms have passed at minimum,
+     thus 25ms is the first interval, doubling until a max interval of 100ms is reached. */
+  static gint timeouts[] = {25, 50, 100};
+
+  g_autoptr(GVariant) signal_variant = NULL;
+  BwrapinfoWatcherData *data = user_data;
+  PidData *pid_data;
+  guint pid;
+  int child_pid;
+  int relative_child_pid = 0;
+
+  pid = data->pid;
+  child_pid = flatpak_instance_get_child_pid (data->instance);
+  if (child_pid == 0)
+    {
+      g_debug ("Failed to read child PID, trying again in %d ms", timeouts[data->timeout_index]);
+
+      if (data->timeout_index == G_N_ELEMENTS (timeouts))
+        /* Already at the max timeout, no need to change. */
+        return G_SOURCE_CONTINUE;
+
+      /* Re-add with the next timeout to use. */
+      g_timeout_add (timeouts[data->timeout_index++], check_child_pid_status, data);
+      return G_SOURCE_REMOVE;
+    }
+
+  g_object_unref (data->instance);
+  g_free (data);
+
+  pid_data = g_hash_table_lookup (client_pid_data_hash, GUINT_TO_POINTER (pid));
+
+  /* Process likely already exited if pid_data == NULL, so don't send the
+     signal to avoid an awkward out-of-order SpawnExited -> SpawnStarted. */
+  if (pid_data == NULL)
+    {
+      g_warning ("%u already exited, skipping SpawnStarted", pid);
+      return G_SOURCE_REMOVE;
+    }
+
+  /* Only send the child PID if it's exposed */
+  if (pid_data->expose_pids)
+    {
+      g_autoptr(GError) error = NULL;
+      relative_child_pid = get_child_pid_relative_to_parent_sandbox (child_pid, &error);
+      if (relative_child_pid == 0)
+        g_error ("Failed to find relative PID for %d: %s", child_pid, error->message);
```

Comment:
g_error is a fatal operation that will kill the portal process, i don't thnk we want that.